src/ProductManifold.jl

```diff
@@ -0,0 +1,416 @@
+eltype(x::ProductMPoint) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductMPoint) = ProductMPoint(map(similar, x.parts)...)
+similar(x::ProductMPoint, ::Type{T}) where T = ProductMPoint(map(t -> similar(t, T), x.parts)...)
+
+function proj_product(x::ProductMPoint, i::Integer)
+    return x.parts[i]
+end
+
+"""
+    ProductTVector(parts)
+
+A more general but slower representation of tangent vectors on a product
+manifold.
+"""
+struct ProductTVector{TM<:Tuple} <: TVector
+    parts::TM
+end
+
+ProductTVector(vectors...) = ProductTVector{typeof(vectors)}(vectors)
+eltype(x::ProductTVector) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductTVector) = ProductTVector(map(similar, x.parts)...)
+similar(x::ProductTVector, ::Type{T}) where T = ProductTVector(map(t -> similar(t, T), x.parts)...)
+
+(+)(v1::ProductTVector, v2::ProductTVector) = ProductTVector(map(+, v1.parts, v2.parts)...)
+(-)(v1::ProductTVector, v2::ProductTVector) = ProductTVector(map(-, v1.parts, v2.parts)...)
+(-)(v::ProductTVector) = ProductTVector(map(-, v.parts))
+(*)(a::Number, v::ProductTVector) = ProductTVector(map(t -> a*t, v.parts))
+
+function proj_product(x::ProductTVector, i::Integer)
+    return x.parts[i]
+end
+
+"""
+    ProductCoTVector(parts)
+
+A more general but slower representation of cotangent vectors on a product
+manifold.
+"""
+struct ProductCoTVector{TM<:Tuple} <: CoTVector
+    parts::TM
+end
+
+ProductCoTVector(covectors...) = ProductCoTVector{typeof(covectors)}(covectors)
+eltype(x::ProductCoTVector) = eltype(Tuple{map(eltype, x.parts)...})
+similar(x::ProductCoTVector) = ProductCoTVector(map(similar, x.parts)...)
+similar(x::ProductCoTVector, ::Type{T}) where T = ProductCoTVector(map(t -> similar(t, T), x.parts)...)
+
+function proj_product(x::ProductCoTVector, i::Integer)
+    return x.parts[i]
+end
+
+function isapprox(M::ProductManifold, x, y; kwargs...)
+    return all(t -> isapprox(t...; kwargs...), ziptuples(M.manifolds, x.parts, y.parts))
+end
+
+function isapprox(M::ProductManifold, x, v, w; kwargs...)
+    return all(t -> isapprox(t...; kwargs...), ziptuples(M.manifolds, x.parts, v.parts, w.parts))
+end
+
+function representation_size(M::ProductManifold, ::Type{T}) where {T}
+    return (mapreduce(m -> sum(representation_size(m, T)), +, M.manifolds),)
+end
+
+manifold_dimension(M::ProductManifold) = sum(map(m -> manifold_dimension(m), M.manifolds))
+
+struct ProductMetric <: Metric end
+
+function det_local_metric(M::MetricManifold{ProductManifold, ProductMetric}, x::ProductArray)
+    dets = map(det_local_metric, M.manifolds, x.parts)
+    return prod(dets)
+end
+
+function inner(M::ProductManifold, x, v, w)
+    subproducts = map(inner, M.manifolds, x.parts, v.parts, w.parts)
+    return sum(subproducts)
+end
+
+function exp!(M::ProductManifold, y, x, v)
+    map(exp!, M.manifolds, y.parts, x.parts, v.parts)
+    return y
+end
+
+function exp(M::ProductManifold, x::ProductMPoint, v::ProductTVector)
```

Comment:
I wonder how the efficiency of this approach compares to creating an empty `ProductMPoint` and then filling it using `exp!`, since (I think) the approach here creates a bunch of mutable points using `similar` and then copies those points to a nw array within the `ProductMPoint`.

Reply:
The points are not copied to make `ProductMPoint`, there are just new references to the points allocated by `exp` on specific manifolds.